<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Error Handling</title>
</head>
<body>
    <script>
        /*try{
            var x = mul(3,4);
        }
        catch(e){
            console.log(e);
        }
        finally{
            console.log("This is final result")
        }


        try{
            //throw "There was an error";     //They both are same
            //throw ("There was an error");
        }
        catch(e){
            console.log(e);
        }

        // Error through new constructor
        try{
            //throw Error("There was an error");
            throw new Error("There was an error while downloading");
                
        }
        catch(e){
            console.log(e); // print the whole information
            console.log(e.name);    //print only name that is "Error"
            console.log(e.message);//print only message
            console.log(e.stack); // print the whole information work like console.log(e);
        }
        //Error message with error code
        try{
            throw{
                a : 202, b : "Failed Execution"
            }
        }
        catch(e){
            console.log( e.a +"-"+e.b)
        }

        // Error Handling by Example
          function xyz(){
            try{
               var x = 10, y = 20, z=40;
                console.log("The sum of x y z is:"+ (x + y + z));
            }
            catch(e){
                console.log(e);
                console.log(e.message);                
            }
            finally{
                console.log("Finally,block always execute!");
            }
        }        
        xyz();

        //second example  on error handling
        function abc(){
            var a = 10, b = -20, c = 5;
            try{
                if(a < 0 || b < 0 || c < 0){
                    throw("Error of Negative value")
                }else{
                    console.log(Math.sqrt(a));
                    console.log(Math.sqrt(b));
                    console.log(Math.sqrt(c));
                 }
            }
            catch(e){
                console.log(e);
                console.log(e.message);                
            }
           finally{
                console.log(Math.sqrt(a));
                console.log(Math.sqrt(b));
                console.log(Math.sqrt(c));
            }
        }        
        abc();

        //Error() function

        let a = [1,3,4,5,6,7,11,23];
        console.log(a);

        let b = [1,3,4,5,6,7,11,23];
        console.error(b);

        console.log("-------------------");
        let obj = {Name:"simmi",age:28};
        console.log(obj);

        let obj1 = {Name:"simmi",age:28};
        console.error(obj1);

        // isNan() function understanding
        function mul(a,b){
            if(isNaN(a) || isNaN(b)){
                return "NaN number is not allowed";
            }
            return a * b;
        }
        console.log(mul(2,4));
        //console.log(mul(2,t));*/

        function divide(a,b){
            if(isNaN(a) || isNaN(b)){
                throw "A value suppiled not a number!";
            }
        }
            try{
                divide("t",2);
            }        
        catch(e){
            console.error(e);
        }
        


    </script>
</body>
</html>